---
title: "BCHM421_FinalProject0"
author: "Group C"
date: "2025-12-03"
output:
  html_document:
  runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(vroom)
library(shiny)
library(ggplot2)
library(magrittr)
library(DT)
library(bslib)
library(htmltools)
library(pheatmap)
library(ggsci)
load("ObjectSave.rda")
```

## Make widget for RNAseq data table
```{r, echo=FALSE}

actionButton("show_RNAseq_table_help", "?")

observeEvent(input$show_RNAseq_table_help, {
  showModal(
    modalDialog(
      title = "RNA Seq data table",
      easyClose = TRUE,
tags$p("This table presents the complete differential gene expression results. Users can search for specific gene symbols using the search bar or use the column filters to identify genes meeting specific statistical thresholds (e.g., FDR < 0.05)."),
      tags$p("Key columns include:"),
      tags$ul(
        tags$li(tags$b("logFC (Log2 Fold Change):"), " Represents the magnitude and direction of gene expression change. Positive values indicate upregulation, while negative values indicate downregulation compared to the control."),
        tags$li(tags$b("logCPM (Log Counts Per Million):"), " Indicates the average expression level (abundance) of the gene across the samples."),
        tags$li(tags$b("PValue & FDR:"), " Statistical measures of significance. The FDR (False Discovery Rate) is the P-value adjusted for multiple testing; values below 0.05 are typically considered significant."),
        tags$li(tags$b("Silenced:"), " Indicates which gene (PRMT5 or WDR77) was knocked down in the experiment.")
      ),
      tags$p("The entire dataset or filtered subsets can be exported using the 'Download' button as either a CSV or Excel file.")
    )
  )
})

renderDataTable({
  datatable(RNAseq,
            filter = list(position = "top", clear = FALSE, plain = TRUE),
            extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           pageLength = 10,
                           buttons = list(extend='collection',
                                          buttons=c('csv', 'excel'),
                                          text='Download'))
  ) %>%
    formatSignif(columns=c('logFC', 'logCPM', 'PValue', 'FDR'), digits=4)
})
```

## Make multi-RNAseq symbol select graph
```{r, echo = FALSE}
inputPanel(
  selectInput(
    inputId = "symbol",
    label = "Select Symbol(s):",
    multiple = TRUE,
    choices = unique(RNAseq$SYMBOL)
  ),
  actionButton("show_RNAseq_plot_help_symbol", "?"),
  downloadButton("download_symbol_plot", "Download Plot")
)

observeEvent(input$show_RNAseq_plot_help_symbol, {
  showModal(
    modalDialog(
      title = "Column plot charting logFC values for select genes based on the RNAseq data table.",
      easyClose = TRUE,
      tags$p("The following column plot displays the log of the fold-change of the expression of our genes of interest, PRMT5 and WDR77. Key details include:"),
      tags$ul(
        tags$li(tags$b("Log Fold-Change:"), " e.g., a 2-fold change is represented by +1 on the y-axis, whereas a 0.5-fold change, meaning expression is halved, is represented by -1 on the y-axis."),
        tags$li(tags$b("Symbol Selector:"), " The symbol selected by the user via the dropdown box represents the gene that was silenced for that given trial."),
        tags$li(tags$b("Color Coding:"), " Upon selecting this gene, the user can see both the change in expression of the PRMT5 gene (represented by the pink column) and the change in expression of the WDR77 gene (represented by the blue column) from baseline.")
      )
    )
  )
})

symbol_plot_reactive <- reactive({
  req(input$symbol)
  
  symbol_plot <- RNAseq %>%
    filter(SYMBOL %in% input$symbol) %>%
    ggplot(aes(x = SYMBOL, y = logFC)) +
    geom_col(aes(fill = silenced),
             position = "dodge"
    ) +
    labs(
      title = "Results for Select Gene",
      y = "logFC",
      x = "SYMBOL"
    ) +
    theme_gray()

  return(symbol_plot)
})

renderPlot({
  symbol_plot_reactive()
})

output$download_symbol_plot <- downloadHandler(
  filename = function() {
    name_suffix <- if(length(input$symbol) > 1) "Multi_Gene" else input$symbol[1]
    paste("Gene_Expression_", name_suffix, ".png", sep = "")
  },
  content = function(file) {
    ggsave(file, plot = symbol_plot_reactive(), device = "png", width = 10, height = 6)
  }
)
```

## Format GO tibble for widget creation

```{r, echo=FALSE}
GO_widget <- GO_gene %>%
  rename(GO_ID = `GO Term`) %>%
  rename(FDR = `FDR q-value`) %>%
  select(GO_ID, Group, Description, FDR, Enrichment)

GO_widget
```

## GO data table
```{r, echo=FALSE}

actionButton("show_GO_ID_table_help", "?")

observeEvent(input$show_GO_ID_table_help, {
  showModal(
    modalDialog(
      title = "Gene Ontology Data Table",
      easyClose = TRUE,
      tags$p("The user can interact with this data via the search bar and filter drop downs, allowing them to type or select any variable. Users can also select how many entries to display (up to 100) and export the table as .csv or Excel files."),
      tags$p("Key columns include:"),
      tags$ul(
        tags$li(tags$b("GO_ID:"), " Refers to the Gene Ontology identification."),
        tags$li(tags$b("Group:"), " Indicates whether a gene is up- or down-regulated by PRMT5 or WDR77."),
        tags$li(tags$b("Description:"), " Summarizes the biological process, molecular function, or cellular component."),
        tags$li(tags$b("FDR:"), " The adjusted p-value estimating the proportion of false positives."),
        tags$li(tags$b("Enrichment:"), " Indicates the fold-enrichment value compared to chance.")
      )
    )
  )
})

```

```{r, echo=FALSE}
inputPanel(selectInput("filterGO_ID", "Select GO ID:",
            choices = c("All", unique(GO_widget$GO_ID)),
            selected = "All"),
            selectInput("filterGroup", "Select Group:",
            choices = c("All", unique(GO_widget$Group)),
            selected = "All"))
```

```{r, echo=FALSE}
filteredData <- reactive({
  data <- GO_widget
  if (input$filterGroup != "All")
    {data <- subset(data, Group == input$filterGroup)}
  if (input$filterGO_ID != "All") 
    {data <- subset(data, GO_ID == input$filterGO_ID)}
  data})
```

```{r, echo=FALSE}
renderDataTable({
  data <- filteredData()
   if (nrow(data) == 0) {
    return(datatable(data.frame(Message = "No data for this selection")))}
  pal_enrichment <- pal_frontiers("default")(5)
  pal_FDR <- pal_frontiers("default")(5)
  cuts1 <- quantile(data$Enrichment, probs = seq(0, 1, length.out = length(pal_enrichment)))[-1]
  cuts2 <- quantile(data$FDR, probs = seq(0, 1, length.out = length(pal_FDR)))[-1]
datatable(data,
          filter = "none",
          extensions = "Buttons",
          options = list(dom = "Blfrtip", pageLength = 10, buttons = list(extend = "collection",
                                                                          buttons = c("csv","excel"),
                                                                          text = "Download"))) %>%
  formatStyle("Enrichment", backgroundColor = styleInterval(cuts1, values = rev(pal_enrichment))) %>%
  formatStyle("FDR", backgroundColor = styleInterval(cuts2, values = rev(pal_FDR))) %>% 
  formatSignif(columns=c("FDR","Enrichment"), digits=4)})
```

## GO ID Gene Expression Plot

```{r, echo = FALSE}

selectInput("go_selector", 
            "Select a GO ID (start typing to find GO ID)", 
            choices = unique(GO_gene$`GO Term`), 
            multiple = FALSE, 
            selected = "GO:0000041")

inputPanel(
  sliderInput(inputId = "max_genes", 
              label = "Choose maximum number of genes.", 
              min = 1, 
              max = 50, 
              value = 10),
  actionButton("show_GO_ID_plot_help", "?"),
  downloadButton("download_column_plot", "Download Plot")
)

observeEvent(input$show_GO_ID_plot_help, {
  showModal(
    modalDialog(
      title = "Column plot charting logFC values for select genes based on the GO ID.",
      easyClose = TRUE,
      tags$p("The following column plot displays the log of the fold-change of the expression of our genes of interest, PRMT5 and WDR77. Key details include:"),
      tags$ul(
        tags$li(tags$b("Log Fold-Change:"), " e.g., a 2-fold change is represented by +1 on the y-axis, whereas a 0.5-fold change (halved expression) is represented by -1."),
        tags$li(tags$b("Gene Selection:"), " The genes displayed are determined by the selected GO ID."),
        tags$li(tags$b("Color Coding:"), " Pink columns represent the change in expression when PRMT5 is silenced; Blue columns represent the change when WDR77 is silenced.")
      )
    )
  )
})

plot_reactive <- reactive({

  req(input$go_selector)

  target_symbols <- geneIndex %>% 
    filter(GO_ID == input$go_selector) %>% 
    pull(Symbol) %>% 
    unique()

  plot_data <- RNAseq %>% 
    filter(SYMBOL %in% target_symbols)

  top_genes <- plot_data %>%
    group_by(SYMBOL) %>%
    summarize(max_change = max(abs(logFC))) %>%
    arrange(desc(max_change)) %>%
    slice_head(n = input$max_genes) %>%
    pull(SYMBOL)

  GO_column <- plot_data %>%
    filter(SYMBOL %in% top_genes) %>%
    ggplot(aes(x = reorder(SYMBOL, SYMBOL), y = logFC, fill = silenced)) +
    geom_col(position = "dodge") + 
    coord_flip() +
    labs(title = paste("Results for Top", input$max_genes, "Genes for", input$go_selector),
         x = "SYMBOL", 
         y = "logFC") +
    scale_fill_hue(labels = c("PRMT5", "WDR77")) +
    theme_gray() +
    theme(text = element_text(size = 14))
  
  return(GO_column)
})


renderPlot({
  plot_reactive()
})

output$download_column_plot <- downloadHandler(
  filename = function() {
    clean_name <- gsub(":", "_", input$go_selector)
    paste("GO_Analysis_", clean_name, ".png", sep = "")
  },
  content = function(file){
    ggsave(file, plot = plot_reactive(), device = "png", width = 14, height = 8)
  }
)


```

## Heatmap widget

```{r, echo=FALSE}
go_to_symbols <- mep50_geneIndex %>%
  rbind(prmt5_geneIndex) %>%
  select(-grouping)

go_to_symbols
```

```{r, echo=FALSE}
GOIDs <- go_to_symbols %>%
  distinct(GO_ID) %>%
  pull(GO_ID)
```

```{r, echo=FALSE}
cpm_mat <- as.matrix(cpm)
```

```{r, echo=FALSE}
inputPanel(
  selectInput("GOID", "Select one or more GOIDs. Start typing...", choices = GOIDs, multiple = TRUE, selected = "GO:0000041", selectize = TRUE),
  selectInput("clSam", "Cluster by Sample?", choices = c("TRUE", "FALSE"), selected = "FALSE", selectize = TRUE),
  selectInput("clGene", "Cluster by Gene?", choices = c("TRUE", "FALSE"), selected = "FALSE", selectize = TRUE),
  selectInput("distanceSam", "Distance Method for Samples", choices = c("euclidean", "manhattan"), selected = "euclidean", selectize = TRUE),
  selectInput("distanceGene", "Distance Method for Genes", choices = c("euclidean", "manhattan"), selected = "euclidean", selectize = TRUE),
  selectInput("linkage", "Linkage Method for All Clustering", choices = c("single", "complete", "average"), selected= "average", selectize = TRUE),
  selectInput("scale", "Method of Data Scaling", choices = c("none", "row", "column"), selected = "none", selectize = TRUE),
  
  actionButton("show_heatmap_help", "?"),
  downloadButton("download_heatmap", "Download Heatmap")
)


observeEvent(input$show_heatmap_help, {
  showModal(
    modalDialog(
      title = "Interactive Heatmap of Gene Expression",
      easyClose = TRUE,
      tags$p("This heatmap visualizes the Count Per Million (CPM) expression data for genes associated with the selected Gene Ontology (GO) term(s). Color intensity represents the expression level, allowing for quick identification of highly or lowly expressed genes."),
      tags$p("Key features include:"),
      tags$ul(
        tags$li(tags$b("Clustering:"), " Users can toggle clustering for Genes (rows) and Samples (columns) to group similar expression patterns using selected Distance and Linkage methods."),
        tags$li(tags$b("Scaling by Row (Gene):"), " Highlights relative changes (up/down-regulation) for a specific gene across samples, ignoring absolute abundance."),
        tags$li(tags$b("Scaling by Column:"), " Normalizes differences between sample libraries.")
      )
    )
  )
})

```

```{r, echo=FALSE}

heatmap_data <- reactive({
  req(input$GOID, input$distanceSam, input$distanceGene, input$linkage, input$scale)
  
  selected_genes <- go_to_symbols %>%
    filter(GO_ID %in% input$GOID) %>%
    pull(Symbol) %>%
    unique()
    
  cpm_sub <- cpm_mat[cpm_mat[,1] %in% selected_genes, ,drop = FALSE]
  
  cpm_numeric <- apply(cpm_sub[,-1,drop=FALSE], 2, function(x) as.numeric(trimws(x)))
  rownames(cpm_numeric) <- cpm_sub[,1]
  
  return(cpm_numeric)
})

renderPlot({
  pheatmap(heatmap_data(), 
           scale = input$scale, 
           cluster_rows = as.logical(input$clGene), 
           cluster_cols = as.logical(input$clSam), 
           clustering_distance_rows = input$distanceGene, 
           clustering_distance_cols = input$distanceSam,
           clustering_method = input$linkage)
}, 
res = 150, 
height = function(){
  max(1200, nrow(heatmap_data()))
}, 
width = 1200
)



output$download_heatmap <- downloadHandler(
  filename = function() {
    paste("Heatmap_", paste(input$GOID, collapse="_"), ".png", sep = "")
  },
  content = function(file) {
    file_height <- max(1200, nrow(heatmap_data()) * 12)
    png(file, width = 1200, height = file_height, res = 150)
    pheatmap(heatmap_data(), 
             scale = input$scale, 
             cluster_rows = as.logical(input$clGene), 
             cluster_cols = as.logical(input$clSam), 
             clustering_distance_rows = input$distanceGene, 
             clustering_distance_cols = input$distanceSam,
             clustering_method = input$linkage)
    
    dev.off()
  }
)

```

